Source/DafnyCore/AST/Formatting.dfy

```diff
@@ -0,0 +1,157 @@
+  }
+}
+module {:extern "Microsoft"} {:options "-functionSyntax:4"}  Microsoft {
+  module {:extern "Dafny"} Dafny {
+    module {:extern "Formatting"} Formatting {
+      import opened MicrosoftDafny
+      import opened System
+
+      const {:extern "System", "String.Empty"} CsStringEmpty: CsString;
+
+      trait IIndentationFormatter {
+        function Reindent(token: IToken, trailingTrivia: bool, precededByNewline: bool, indentation: CsString, lastIndentation: CsString): CsString
+        // Returns -1 if no indentation is set
+        method GetIndentation(token: IToken, currentIndentation: CsString)
+          returns (
+            indentationBefore: CsString,
+            indentationBeforeSet: bool,
+            lastIndentation: CsString,
+            lastIndentationSet: bool,
+            indentationAfter: CsString,
+            indentationAfterSet: bool)
+          requires token.Valid()
+          ensures token.allTokens == old(token.allTokens)
+      }
+
+      lemma IsAllocated<T>(x: T)
+        ensures allocated(x) {
+      }
+
+      /** Prints the entire program while fixing identation, based on a map */
+      method printSourceReindent(firstToken: IToken, reindent: IIndentationFormatter) returns (s: CsString)
+        requires firstToken.Valid()
+        ensures forall token <- firstToken.allTokens :: s.Contains(token.val)
+      {
+        var token: IToken? := firstToken;
+        s := CsStringEmpty;
+        var currentIndent := CsStringEmpty;
+        ghost var i := 0;
+        var leadingTriviaWasPreceededByNewline := true;
+        var allTokens := firstToken.allTokens;
+        while(token != null)
+          decreases if token == null then 0 else |token.allTokens|
+          invariant 0 <= i <= |allTokens|
+          invariant if token == null then i == |allTokens| else
+                    && token.Valid()
+                    && i < |allTokens|
+                    && token == allTokens[i]
+          invariant forall t <- allTokens[0..i] :: s.Contains(t.val)
+        {
+          if(token.Next == null) {
+            firstToken.TokenNextIsNullImpliesLastToken(token, i);
+          } else {
+            firstToken.TokenNextIsIPlus1(token, i);
+          }
+          IsAllocated(allTokens[0..i]);
+          var indentationBefore, indentationBeforeSet,
+              lastIndentation, lastIndentationSet,
+              indentationAfter, indentationAfterSet := reindent.GetIndentation(token, currentIndent);
+
+          var newLeadingTrivia := reindent.Reindent(token, false, leadingTriviaWasPreceededByNewline, indentationBefore, lastIndentation);
+          var newTrailingTrivia := reindent.Reindent(token, true, false, indentationAfter, indentationAfter);
+          leadingTriviaWasPreceededByNewline := HelperString.FinishesByNewline(token.TrailingTrivia);
+          ghost var sPrev := s;
+          s := String.Concat(s, String.Concat(newLeadingTrivia, String.Concat(token.val, newTrailingTrivia)));
```

Comment:
Shouldn't you use a StringBuilder here, to get linear instead of quadratic performance?

Reply:
I thought that C# would be clever, but you're right, better be safe than sorry !